### DISCUSSION TITLE
Performance question

### DISCUSSION THREAD

**Discussion by Ajay100000** [dotnet/runtime#85306] [General] [answered]
Since I don't now who to ask, hopefully someone sees this who can carry this forward (and understand IL code at an expert level).

There is a Microsoft employee who has written a brilliant bit of code in terms of performance, for deep cloning. He has stopped work on it, perhaps someone else at Microsoft can pick it up and give c# this functionality.

The github repository is:

https://github.com/ReubenBond/DeepCopy

If a performance evaluation is done on this code, it is extremely fast and efficient. This is a benchmark I have, as an example:

|                             Method |            Mean |         Error |        StdDev |          Median |
|----------------------------------- |----------------:|--------------:|--------------:|----------------:|
|                ICloneable_Benchmark |        18.70 ns |      0.426 ns |      0.597 ns |        18.87 ns |
|           DeepCopyLibrary_Benchmark |        46.07 ns |      0.877 ns |      1.997 ns |        46.78 ns |
|                Reflection_Benchmark |       831.06 ns |     15.782 ns |     33.290 ns |       847.12 ns |
|         JSONSerialization_Benchmark |     1,644.78 ns |     32.694 ns |     61.408 ns |     1,674.52 ns |

Perhaps an IL expert at Microsoft can keep this going and release it as a performance improvement for c#. I know that I and many other are using less efficient deepcopy/deepclone algorithms and would be glad to use something this efficient (especially for games development).

Anyway, if possible to bring this kind of efficiency in c#, then I would be very grateful.

**Comment by huoyaoyuan** [comment]
General discussion of a library feature can be posted at dotnet/runtime.

> If a performance evaluation is done on this code, it is extremely fast and efficient.

There's no magic behind it. It just generates the same IL matches what you write to copy field by field.

> Perhaps an IL expert at Microsoft can keep this going and release it as a performance improvement for c#.

Copying arbitrary fields from any type of objects not authored by you can be security concern. You can also see that `ICloneable` is faster in the benchmark because the object can control and optimize the implementation. Reflection and serialization are really not comparable here because their approaches are completely different.

**Comment by cincuranet** [comment]
Transferring to _runtime_.

**Comment by stephentoub** [comment]
cc: @ReubenBond

**Comment by Ajay100000** [comment]
For reference, doing the IClonable way is a lot of work. For safety, I am using the JSONSerialization for deep cloning as I do not have the knowledge of IL. However, if the IL way was the new standard for deep cloning, a lot of developers would use it due to the massive speed increase.

Thanks for making this a discussion.

**Reply by ReubenBond** [reply]
I made that library mostly for demonstration and experimentation and I don't see it as a viable library for production use.

Orleans includes a facility for cloning objects based on a C# Source Generator, requiring you to add attributes to your types and properties. It might be viable for your use case.

**Comment by Ajay100000** [comment, marked as answer]
Thanks for the information, that library was very impressive.

I raised this issue as I thought it would help with the general C# push towards improved performance. I will stick with the JsonSerialization way as most of my use cases do not need extreme performance and therefore JsonSerialization is good enough and easy to reason as it is only a couple of lines of code (In my programming style, generally like POCO objects rather than ones that need attributes).

I was really impressed with what you had done which is why I thought someone with expert IL knowledge would want to provide this kind of speed improvement to the general C# infrastructure. My motivation is for C# to have the best performance it can.
